AC-552 word cloud accessibility updates

clrux opened the pull request:
# [AC-552](https://openedx.atlassian.net/browse/AC-552)

This updates our very popular word cloud advanced module to be accessible. This has been achieved by the following:

* adding a new field in the xblock editor window for "Instructions" which allows course authors to add instructions for the word cloud (e.g., "List your five favorite foods.")
* added various unique `id`'s to the individual SVG words, and 
* linked those to the overall SVG word with `aria-describedby`
* added `label`'s to the text inputs for each word

Our word cloud is very popular among courses, but lacked a lot of very important elements to make it accessible.

## Tests

I opted not to include tests since this is all programatic and unconfigurable.

## Sandbox

Studio: https://studio-clrux-ac-552.sandbox.edx.org/container/block-v1:edX+DemoX+Demo_Course+type@vertical+block@45c7cedb4bfe46f4a68c78787151cfb5

LMS: https://clrux-ac-552.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/courseware/9fca584977d04885bc911ea76a9ef29e/07bc32474380492cb34f76e5f9d9a135/1?activate_block_id=block-v1%3AedX%2BDemoX%2BDemo_Course%2Btype%40vertical%2Bblock%403785b1366e7749edacf28d5f7dc6f98a

## Reviewers

- [x] @cptvitamin (Accessibility)
- [x] @cahrens (T&L)
- [x] @sstack22 (Product)

clrux commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -126,7 +129,7 @@ define('WordCloudMain', [], function () {
         scaleFactor = 1;
         maxFontSize = 200;
-        minFontSize = 15;
+        minFontSize = 16;
```
;-)

catong commented on common/lib/xmodule/xmodule/word_cloud_module.py:
```diff
@@ -41,6 +41,12 @@ class WordCloudFields(object):
+    )
     num_inputs = Integer(
         display_name=_("Inputs"),
         help=_("Number of text boxes available for students to input words/sentences."),
```
@clrux Any chance we can sneak in a terminology fix in this PR? "students" -> "learners". Thanks!

clrux replied:
Sure!

catong commented on common/lib/xmodule/xmodule/word_cloud_module.py:
```diff
@@ -41,6 +41,12 @@ class WordCloudFields(object):
     )
+    instructions = String(
+        display_name=_("Instructions"),
+        help=_("Add instructions for this word cloud"),
```
@clrux Is there a char limit for the instructions?

clrux replied:
I don't think so, but I'd keep it short for presentation.

sstack22 commented:
So pumped! Looks good 👍 . One nit - ideally Instructions would be above Inputs.

clrux commented:
@sstack22 Yeah... it's output alphabetically. You think it's annoying enough for me to try and figure it out?

sstack22 commented:
@clrux - Ah, I see. No worries, I don't think it's high priority.

cahrens commented:
The relative size of the text (instructions vs. my words) is quite odd. Can the UX be improved some?

![image](https://cloud.githubusercontent.com/assets/484484/17380105/61fd0918-5993-11e6-91b7-68751bab94cb.png)

cahrens commented on common/lib/xmodule/xmodule/word_cloud_module.py:
```diff
@@ -41,9 +41,15 @@ class WordCloudFields(object):
+    instructions = String(
+        display_name=_("Instructions"),
+        help=_("Add instructions for this word cloud"),
+        scope=Scope.settings,
```
It feels like this should really be Scope.content (though perhaps it is arguably the same as display name). In general the line between Scope.content and Scope.settings is very murky, so I'm OK with all of these being Scope.settings.

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -230,15 +233,30 @@ define('WordCloudMain', [], function () {
             var percent = (response.display_student_percents) ? ' ' + (Math.round(100 * (stat / response.total_count))) + '%' : '';
 
-            studentWordsKeys.push('<strong>' + word + '</strong>' + percent);
+            studentWordsKeys.push(HtmlUtils.interpolateHtml(
```
You can simplify this code by using HtmlUtils.joinHtml. For example, 

````
var htmlSnippet = HtmlUtils.joinHtml.apply(
                this, _.map(grouped, function(options, facetKey) {
                    if (options.length > 0) {
                        return this.renderFacet(facetKey, options);
                    }
                }, this)
            );

cahrens replied:
You can ignore this comment-- see my comment above/below.

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -230,15 +233,30 @@ define('WordCloudMain', [], function () {
+                }
+            ));
         });
         studentWordsStr = '' + studentWordsKeys.join(', ');
```
Wouldn't be necessary with HtmlUtils.joinHtml.

clrux replied:
What do you mean? This is pushing a new object into the `studentWordsKeys` array.

cahrens replied:
What I meant was that you could replace the "$.each(response.student_words, function (word, stat)" loop and studentWordsKeys with a usage of HtmlUtils.joinHtml. But looking at this further, it wouldn't simplify the code much because you still want to deal with putting commas in-between the individual words (which can have internationalization impact, but that's an existing bug).

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -230,15 +233,30 @@ define('WordCloudMain', [], function () {
+
+        HtmlUtils.setHtml(
+            cloudSectionEl.find('.your_words'),
+            HtmlUtils.HTML(studentWordsStr)
```
There would be no need to wrap HtmlUtils.HTML around the response from HtmlUtils.joinHTML, as it would already be an HTML snippet.

cahrens replied:
You can ignore this comment (see my note above).

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -230,15 +233,30 @@ define('WordCloudMain', [], function () {
+
+        HtmlUtils.setHtml(
+            cloudSectionEl.find('.your_words').end().find('.total_num_words'),
+            HtmlUtils.HTML(response.total_count)
```
This isn't actually HTML, it's just a number. I think it should actually be escaped.

Instead of using setHtml, just use .text (it is a span element).

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -250,10 +268,17 @@ define('WordCloudMain', [], function () {
-                .enter().append('g');
+                .enter()
+                .append('g')
+                .attr('aria-labelledby', function(d) {
```
This is invalid. Also make sure that IDs are unique (even with multiple word clouds on the same page, possibly with the same words).

![image](https://cloud.githubusercontent.com/assets/484484/17380637/1da787b8-5996-11e6-9363-f730fb6f447c.png)

clrux replied:
Ah, I need to strip spaces.

clrux replied:
They should be unique though, since I'm using the word cloud ID in the individual ID.

cahrens commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,38 @@
+
+    <div class="input_cloud_section" role="group">
         % for row in range(num_inputs):
+            <label for="word_cloud_${element_id}_input_${row}" class="sr">${_('Word')} ${row + 1} ${_('of')} ${num_inputs}</label>
```
You can't do concatenation like this and have any prayer of it translating properly. Use placeholders.

clrux replied:
I simplified with "1/3". Don't need to get too fancy.

cahrens commented:
@clrux Why not just use the Word Cloud display name (which is not currently rendered in the LMS) instead of introducing a new instructions field?

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -230,15 +233,30 @@ define('WordCloudMain', [], function () {
-            studentWordsKeys.push('<strong>' + word + '</strong>' + percent);
+            studentWordsKeys.push(HtmlUtils.interpolateHtml(
+                gettext('{startTag}{word}{endTag}{percent}'),
+                {
```
Remove the gettext usage here. There is nothing to translate-- these are all placeholders, so the translators will be very confused. Plus, we want the words to appear in whatever language they were entered in.

cahrens commented:
@clrux @sstack22 Thinking more about this, I'm wondering if the Display Name should always show (currently it does not), but the instructions should ONLY show before the learner has filled in their words. Maybe we can just catch up in person on this?

clrux commented:
@cahrens A few things:

* I've addressed your feedback, please take another pass.
* Display Name in the module _could_ be used for instructions, but it's not clear it _can_ be. It's the title of the component. I'd prefer to give authors additional room to explain the information they're asking for. Is there a specific reason not to include an additional field?
* I've allowed for words with spaces ("french fries") by replacing the spaces with underscores for use as parameter values.
* I've gone back to `aria-describedby` which allows space-separated values, rather than just a single ID reference. Should be valid now.
* Sandbox should be updated in about 15 minutes.

clrux commented:
jenkins run all

cahrens commented:
@clrux I talked with @sstack22, and she will follow up with how we want to handle display name vs. instructions (looking also at existing usages of word cloud). I'm going to hold off on further reviews until product has had time to weigh in.

sstack22 commented:
@clrux, @cahrens - I've looked at current examples, and course teams are currently using a text component above word cloud to include instructions. Here's an example: https://courses.edx.org/courses/course-v1:WellesleyX+SHKSPR_02X+2T2016/courseware/50176233a6104091a373187d817476ba/faa142630de047d09d93ef381dcae265/

-  I do think we should keep the new Instructions field, and that it is fine for it to persist after you have answered, as this mimics current behavior. 

- @clrux how would this impact existing word cloud components? I would expect for the instructions field to remain blank for them. I just want to make sure "List your five favorite foods" doesn't appear everywhere :)

- We should also display the display name in the LMS. It seems buggy that it does not currently do so.

clrux commented:
@sstack22 The new "instructions" field would be present for existing clouds if they were to edit their cloud, or make a new one. I don't think it will impact older clouds as that data is already stored without the new field.

However, the default text is "List your five favorite foods" and if that isn't cleared out or changed, will show up on new clouds. We would need to update documentation to instruct course authors to either change it and use the field, or clear it out (preferably the former).

I'll update the code to get the display name in there as a heading for each instance. I'll try to get this done today.

cahrens commented:
@clrux We need to test with pre-existing word cloud instances to be sure of the behavior. My guess is that the default display name and instructions will appear, which is not what we want.

To test, create a course with a word cloud instance on staging/master, export the course, and then import it on your sandbox with the change.

clrux commented:
@sstack22 @cahrens I've done the following (and the sandbox is updated):

* added the component display name as a heading (like all the others)
* moved the "total words" and "your words" to the bottom, under the actual rendered cloud
* used a list for each textual word, rather than a comma-separated string
* various UI cleanup

clrux commented:
@cahrens You are correct, it is present on older courses. Should I not set a default value?

cahrens commented:
@clrux I think you need to set a default value, else nobody will ever see it. I question though the particular default value you chose. Perhaps something more like "Replace this with your own instructional text"? Work with product or doc to figure this out.

But as far as functionality, can you only render the instructional text (and display name) if they do not match the default values (without hard-coding default values in more than one location)?

sstack22 commented:
@clrux , @cahrens  - I'm a little confused - is it possible to set a default value without impacting previous courses? I think it is important that no text appear in the LMS unexpectedly.

clrux commented:
@sstack22 Yep, working on it :)

catong commented on lms/djangoapps/courseware/tests/test_word_cloud.py:
```diff
@@ -248,7 +248,9 @@ def test_word_cloud_constructor(self):
             'element_class': self.item_descriptor.location.category,
             'element_id': self.item_descriptor.location.html_id(),
+            'instructions': 'Replace this text with instructions for your word cloud.',
+            'instructions_detault': 'Replace this text with instructions for your word cloud.',
```
@clrux Typo: instructions_detault -> instructions_default

catong commented:
@cahrens, @sstack22 FYI the doc PR associated with these word cloud updates is https://github.com/edx/edx-documentation/pull/1158. I've tagged you both for review there. Thanks!

cahrens commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,41 @@
                 ${'style="display: none;"' if submitted else ''}
                 type="text"
                 size="40"
+                aria-describedby="word_cloud_${element_id}_instructions"
```
Note that this element may not exist.

cahrens commented on common/lib/xmodule/xmodule/word_cloud_module.py:
```diff
@@ -239,7 +245,10 @@ def get_html(self):
+            'submitted': self.submitted,
+            'instructions_default': WordCloudFields.instructions.default,
+            'instructions': self.instructions,
+            'display_name': self.display_name
```
Do we also need to pass the display_name default to avoid it suddenly appearing on existing instances?

clrux replied:
Is that how it works with other components? Do we hide the default name, or hide it if it's unchanged?

cahrens replied:
@clrux No, it's not how we handle it in other components. But in other components, the display name was always shown in the LMS.

My concern here (which I believe is @sstack22's concern as well) as that Word Cloud instances in running courses will suddenly show the default Word Cloud display name, as course teams will not have thought to set it previously.

clrux replied:
OK so I'll add a check to see if it differs from the default, like I did with instructions. That way it won't show up unless it's been changed.

clrux commented:
jenkins run all

clrux commented:
jenkins run lettuce

clrux commented:
@cahrens @sstack22 I've updated it to only show the `display_name` if it's been changed from the default. Sandbox is updated.

cahrens commented on common/lib/xmodule/xmodule/word_cloud_module.py:
```diff
@@ -248,7 +248,8 @@ def get_html(self):
             'instructions': self.instructions,
-            'display_name': self.display_name
+            'display_name': self.display_name,
+            'display_name_default': WordCloudFields.display_name.default
```
Sorry to mention this now, but wouldn't it be easier to not pass the defaults to the HTML, and instead take into account the default values when display_name and instructions are passed?

````
'instructions': if self.instructions is not WordCloudFields.instructions.default self.instructions else None
````

Then the HTML check is simplified to whether the element is None.

clrux replied:
Sure, I'll play with it.

clrux replied:
OK so the only variables available to the underscore template are those passed through context. So, I'd either need to pass the individual defaults (as I was before), or pass in the entire object. I think I'm going to keep it like this for clarity.

cptvitamin commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,45 @@
+    </div>
+    % endif
+
+    <div class="input_cloud_section" role="group">
```
@clrux  This looks like a fieldset/legend situation to me. If you want to emulate a fieldset/legend, the instructions, which serve as the label for the group, would be a child of this div. The div would have an aria-labelledby which referenced the instructions.

cptvitamin commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,45 @@
                 type="text"
                 size="40"
+                % if instructions is not instructions_default:
+                aria-describedby="word_cloud_${element_id}_instructions"
```
If you follow the convention above, this aria-describedby is unnecessary and would get very chatty, especially if the question text is long.

cptvitamin commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,45 @@
 
-    <section id="result_cloud_section_${element_id}" class="result_cloud_section">
-        <h3>${_('Your words:')} <span class="your_words"></span></h3>
-        <h3>${_('Total number of words:')} <span class="total_num_words"></span></h3>
```
you removed some functionality here.  There is the list of words that the student chooses, but then there is the list of words that the entire class chooses.  These are two separate items.  They seem to be mashed into one below.

clrux replied:
I've moved this to the bottom, under the word cloud.

clrux replied:
Ah I see. I changed the context, assuming it was per-user. I've changed this back to be:

"Total are __n__ words in total."
"Your words were:"
...

catong replied:
@clrux Were the original strings "Your words" and "Total number of words"? For backwards compatibility, should we keep them the same?

catong replied:
Thanks @clrux
I'd suggest:
"x words submitted in total."
"Your words were:"

clrux replied:
All set here, thanks!

cptvitamin commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,45 @@
+    <div id="result_cloud_section_${element_id}" class="result_cloud_section">
         <div class="word_cloud"></div>
-    </section>
+        <p id="heading-words">${'Your'} <strong><span class="total_num_words"></span></strong> ${'words'}:</p>
```
In a page with multiple word clouds, this will not be unique.

cptvitamin commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -154,136 +154,166 @@
-                    }
+                .append('title')
+                .attr('id', function(d) {
+                    return 'title_word_' + d.text;
```
you need to replace all of the spaces in d.text with underscores if you want the ID attributes to work.

cptvitamin commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -154,136 +154,166 @@
-
+                .append('text')
+                .attr('id', function(d) {
+                    return 'text_word_' + d.text;
```
again, make sure you replace spaces with underscores.

cptvitamin commented:
RE: Default value for Instructions and legacy support: Can we just make the Instructions field required for newly authored Word Clouds in Studio? That way, older ones won't be affected, but if you create a new one, or edit an old one, you have to add instructions.

clrux commented:
@cptvitamin I don't think there's a way to differentiate between old and new word clouds as both use the same templates. I also think we'd have to compare - set a marker of some sort indicating where the "cutoff" between old and new is, and then have the templates check that when it renders.

cptvitamin commented:
@clrux I was just suggesting a change to studio.  I don't see why we can't leave it as you currently have it (no instruction unless one is specified), but make the new instruction field required so that when you create a new word cloud, or edit an old one, you have to add an instruction.

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -154,136 +154,176 @@
                 var percent = (response.display_student_percents) ? ' ' + (Math.round(100 * (stat / response.total_count))) + '%' : '';
 
-                studentWordsKeys.push('<strong>' + word + '</strong>' + percent);
+                studentWordsKeys.push(HtmlUtils.interpolateHtml(
```
Extreme nit: Now that comma is not being used as a separator, HtmlUtils.joinHtml would work nicely for building up the HTMLSnippet. For an example, see https://github.com/edx/edx-platform/blob/9a367f64fd733cae087d6d0911985364ced01e32/lms/static/js/groups/views/cohort_discussions_course_wide.js#L36

clrux replied:
Does extreme nit mean this is incorrect?

cahrens replied:
"Extreme nit" meant that what you have is correct, but it could be written more succinctly (and perhaps readably) in another way.

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -154,136 +154,176 @@
+                HtmlUtils.HTML(studentWordsStr)
+            );
+
+            cloudSectionEl.find('.your_words').end().find('.total_num_words')
```
The concatenation of response.total_count + " words submitted in total." is not good for translation. A placeholder for the number should be used.

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -154,136 +154,176 @@
+                    .append('g')
+                    .attr('aria-describedby', function(d) {
+                        return HtmlUtils.interpolateHtml(
+                            gettext('text_word_{word} title_word_{word}'),
```
These IDs may not be unique-- in particular if there are 2 word clouds in the same unit that have words in common. You will need to add the xblock ID or something else that is unique.

Also, I wonder about having the user-supplied word be part of the ID. What if it has unicode characters, spaces, or "HTML" in it? Do we need to support HTML 4.x, which does not allow unicode characters in IDs?

clrux replied:
Took me a few hours, but I figured this out. I'm using a unique ID per `g` that is a combination of the word cloud's `id` and a randomly generated number between 1 and 999. This works though the `id`'s are long.

cahrens commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,45 @@
 
-    <section class="input_cloud_section">
         % for row in range(num_inputs):
+            <label for="word_cloud_${element_id}_input_${row}" class="sr">${row + 1}/${num_inputs}</label>
```
Does this read OK for a screen reader? Would something like "Input 1 out of 5" (translated) be clearer?

clrux replied:
I originally had "out of", but it was kinda wordy.

cahrens commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,45 @@
         % for row in range(num_inputs):
+            <label for="word_cloud_${element_id}_input_${row}" class="sr">${row + 1}/${num_inputs}</label>
             <input
+                id="word_cloud_${element_id}_input_${row}"
```
Nest the input inside the label-- then you don't need the id.

clrux replied:
Done!

cahrens commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,45 @@
 
-        <section class="action">
+        <div class="action">
             <input class="save" type="button" value="${_('Save')}" />
```
Shouldn't this be a button instead of an input?

clrux replied:
They [both work](https://jsfiddle.net/yx2e5byd/1/) but `button` is probably cleaner. I'll update.

cahrens commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,45 @@
+    <div id="result_cloud_section_${element_id}" class="result_cloud_section">
         <div class="word_cloud"></div>
-    </section>
+        <p><strong><span class="total_num_words"></span></strong> ${'words submitted in total'}.</p>
```
Again, you shouldn't do this concatenation. Use a placeholder and HTMLUtils.

clrux replied:
Fixed in the JS.

cahrens commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,45 @@
         <div class="word_cloud"></div>
-    </section>
+        <p><strong><span class="total_num_words"></span></strong> ${'words submitted in total'}.</p>
+        <p>${'Your words were'}:</p>
```
This needs to be translated, and include the colon in the translated string, as it won't make sense to be concatenated in all languages.

clrux replied:
Done here.

cahrens commented:
@cptvitamin I don't think the behavior you are describing (enforcing specification of instructions for new instances) is trivial to achieve-- validation would have to be added. @clrux if you wish to try this, see https://github.com/edx/edx-platform/blob/b5ab64505af8d83dfe147771d799af5741056024/common/lib/xmodule/xmodule/library_content_module.py#L501 for an example of using the xblock validation framework (also used by the split_test_module).

clrux commented:
jenkins run all

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -154,136 +158,188 @@
+                    .data(words)
+                    .enter()
+                    .append('g')
+                    .attr('data-id', function() {
```
This seems overly complex. Why can't you just loop over the words and use a simple counter (0, 1, 2, 3...)?

clrux replied:
Not sure that'll work in this case. This has to enter into the SVG and create the `g` with various attributes. It attaches the individual text and title at the same time.

cahrens commented:
Nit: could use a little padding around Save button (and better alignment?).

![image](https://cloud.githubusercontent.com/assets/484484/17751957/9e74a130-6497-11e6-9355-7512c9eb9a7a.png)

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -12,106 +12,110 @@
+        'gettext',
+        'edx-ui-toolkit/js/utils/html-utils'
+    ], function(gettext, HtmlUtils) {
+        function generateUniqueId(wordCloudId, uniqueId) {
```
With the simplification, uniqueId isn't a good name-- it's just a counter.

clrux replied:
This is still a unique ID because it builds a string using the xmodule ID with various other things.

cahrens replied:
I mean that if you remove the random generation of uniqueId and simply use a counter "uniqueId" is no longer unique. "index" would be a more appropriate name.

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -154,136 +158,188 @@
+                .append('title')
+                .attr('id', function() {
+                    return HtmlUtils.interpolateHtml(
+                        gettext('title_word_{word}'),
```
"word" doesn't make sense as a variable name here (now that it is a unique ID).

Same comment down below.

cahrens commented on common/lib/xmodule/xmodule/js/src/word_cloud/word_cloud_main.js:
```diff
@@ -154,136 +158,188 @@
+                    return HtmlUtils.interpolateHtml(
+                        gettext('title_word_{word}'),
+                        {
+                            word: generateUniqueId(cloudSectionEl.attr('id'), $(this).parent().data('id'))
```
I'm surprised that "this" works in this function callback. Happy, but surprised.

cahrens commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,46 @@
-                size="40"
-            />
+            <label>
+                <span class="sr">${row + 1}/${num_inputs}</span>
```
I've asked this before, but how does this actually read with a screen reader?

cahrens commented:
Just a few more things... almost there!

clrux commented on lms/templates/word_cloud.html:
```diff
@@ -1,30 +1,46 @@
+    % endif
+
+    % if instructions is not instructions_default:
+    <div class="input_cloud_section" role="group" aria-labelledby="word_cloud_${element_id}_instructions">
```
It's rendering an additional `aria-describedby` if the instructions were set, otherwise it omits it since the referenced `id` would be empty (and would trigger failures).

cahrens replied:
What I meant is why is the whole div with class "input_cloud_section" rendered if "submitted" is true. In that case a display of none is programmatically added to the div, causing flashing on initial render.

cahrens commented on cms/static/sass/elements/_xblocks.scss:
```diff
@@ -248,6 +248,18 @@
+    .action {
+        
+        .save {
+            @extend %btn-secondary-blue;
```
As we discussed in Hipchat, make sure that:

- This doesn't negatively impact other XBlocks
- It has sufficient color contrast
- It looks good in the LMS (the change here is for CMS only). CMS and LMS should match.

cahrens commented:
👍 Pending save button cleanup.

cptvitamin commented:
@cahrens I didn't know if XBlock validation was so hard to achieve. We could simply use HTML5 validation, which can be achieved simply by adding the ```required``` attribute to the input. Validation gets handled by the user agent.  It's not perfect. If you use an old browser that doesn't support HTML5, then validation won't occur, but it's better than nothing.  As you can see from caniuse.com, this is very well supported in modern browsers:
http://caniuse.com/#search=required

cptvitamin commented:
It seems like you have already resolved the Display Name vs Instructions here, but I wanted to add one more argument for Instructions: The value of Instructions is used as the accessible name for the group, much like legend is the accessible name of a group of radiobuttons.  When we were in the discovery phase of the CAPA work we discussed using the Display Name for the question text, or the accessible label but ultimately decided against it (for good reason).  This would be using the Display Name for the accessible name and I don't think that's right, especially given that many times, author's don't customize Display Name.

cptvitamin commented:
So, i just got a chance to take a look at this on the sandbox. My concern is that we don't do enough to encourage the user to add Instructions. If they use Display name for that, which they are now likely to do since we force them to display it, we don't get an accessible name for the group.  We have an aria-labelledby that references an empty div.  While this passes an automated test, its an accessibility fail.  At the very least, I think we should put a ```required``` attribute on the Instructions field to get free client side validation.  Ideally, what I would do is detect if the Instructions field is blank, and if so, change the aria-labelledby to reference the Display name instead.  That way we always get an accessible name.

clrux commented:
@cptvitamin Adding `required="required"` could work. Users would only encounter it as a required field if they were to edit the xblock and try to save it. It's an easy fix that requires no backend work, and would guarantee us an accessible label. @cahrens is out, but if @sstack22 agrees I'll make this happen tomorrow.

cahrens commented:
I don't see how adding required to the html will have any effect in the Studio editor. The XBlock validation is just an API for presenting messages in Studio-- the XBlock has to do the validation itself.

cahrens commented:
Or is required a property available on XBlock fields?

clrux commented:
After looking at this for several hours today, I don't think this is possible given the existing code. Here's why:

We define the fields for the word cloud in `word_cloud_module.py`. When adding/editing a word cloud, we use the generic `edit_xblock.js` file that, as far as I can tell, is fairly agnostic as to the data it works with. So when saving an xblock (on the front-end) we may have an opportunity to intervene and check for whatever, but those whatevers have to be present. It doesn't look like the script is aware of various attributes (i.e., required).

Currently, if the display name is different from the default, we show it. And currently if the instructions are different from the default, we show them, and they also become the accessible label for the word cloud group. This works if both are set, but unfortunately doesn't solve the problem if neither are set because then the group doesn't have an accessible label.

We need to come up with a solution that guarantees an accessible label, but that doesn't affect existing word clouds. How can we ensure edited or new word clouds get the appropriate information, but older word clouds remain unchanged?

clrux commented:
For information, in today's grooming we discussed doing the following:

* use the display name as the accessible name unless instructions are set, in which case use them
* always show the display name, but only show instructions if they're set
* have documentation encourage course authors to add instructions
* we'll tackle Studio validation later

sstack22 commented:
A final 👍  on my end!

catong commented on common/lib/xmodule/xmodule/word_cloud_module.py:
```diff
@@ -37,25 +37,25 @@ class WordCloudFields(object):
     display_name = String(
         display_name=_("Display Name"),
-        help=_("Display name for this module"),
+        help=_("The label for the word cloud tool on the course page."),
```
@clrux Sorry, could I make one more suggestion for this help text? "The label for this word cloud on the course page."
(it would refer to *this* word cloud, and it would not appear as a "tool")

cptvitamin commented:
Thanks for stickin' with this one. I just tried every permutation I could think of and all were handled gracefully.  👍

catong commented:
Awesome work, @clrux!!

clrux commented:
jenkins run bokchoy
jenkins run python

cahrens commented on lms/djangoapps/courseware/tests/test_word_cloud.py:
```diff
@@ -241,14 +241,16 @@ def test_handle_ajax_incorrect_dispatch(self):
         expected_context = {
             'ajax_url': self.item_descriptor.xmodule_runtime.ajax_url,
+            'display_name': self.item_descriptor.display_name,
+            'display_name_default': 'Word cloud',
```
Please remove this unused variable-- display_name_default.
